### DISCUSSION TITLE
RS485 Expansion Board for XIAO | Seeed Studio Wiki

### DISCUSSION THREAD

**Discussion by giscus** [Seeed-Studio/wiki-documents#1940] [Q&A]
# RS485 Expansion Board for XIAO | Seeed Studio Wiki

XIAO Communcate with RS485 Expansion Board

https://wiki.seeedstudio.com/XIAO-RS485-Expansion-Board/

**Comment by NguyenVanDong94** [comment]
i can convert to rs422?

**Reply by nfs0619** [reply]
Thanks for reaching out! For assistance with converting to RS422, we recommend contacting our technical support team at techsupport@seeed.io. They can provide detailed guidance and clarify any questions regarding your setup.

**Comment by hlima** [comment]
schematics?  when we add a c6 or c3, which pins are connected o RS485?

**Reply by nfs0619** [reply]
For detailed schematics and pin connections when adding C6 or C3 to RS485, please contact our technical support team at techsupport@seeed.io. They can provide accurate documentation and further assistance.

**Reply by hlima** [reply]
Does this board only works when connected to another xiao board? I tried with a device that uses rs485 and it did not work. I set it to input and connected AB pins to the AB pins on the device. I did not use the GROUND pin only AB pins. I am using a XIAO C6 and programmed the serial to use pins D4, D5

**Reply by Harrison-Xu** [reply]
Here is Harrison from Seeed Studio technical support team. 

Sorry we cannot provide the schematic file of this product currently. 

This product cannot work independently without a XIAO connected to it. It's right to use D4 and D5 pins, so please share with us your hardware connection photo and code then we can continue to help you with it.

**Reply by hlima** [reply]
I do have a xiao c6 connected. My problem was that I initially confgured D4 as RX.  Once I corrected that  and configured D5 as RX everything worked. Thank you.

**Reply by Elu43** [reply]
@Harrison-Xu @hlima why do you use D4 or D5 as RX ? Pinout shows D7 (GPIO 17) as RX for the C6 ?

**Comment by hlima** [comment]
can we power the xiao C6  connected to the board using the 12V input (do not use the C6 USBC) ? My RS485 device has a GND/12V output pins and I wanted to use those to power the  whole board including XIAO C6 without having to use an additional USBC power.

**Comment by rickyrockets** [comment]
Where can I find the HardwareSerial library for this?

**Reply by HM-sur** [reply]
![image](https://github.com/user-attachments/assets/b9a72c69-32a8-411b-a7cc-c4c350b839e8)
[The HardwareSerial library](https://github.com/espressif/arduino-esp32/blob/master/cores/esp32/HardwareSerial.h) is in the official esp32 onboard package

**Comment by Elu43** [comment]
Why is there :

```
  mySerial.begin(115200, SERIAL_8N1, 7, 6); // RX=D4(GPIO6), TX=D5(GPIO7)
```

When we can see on the pin diagram that TX=D6=GPIO21 and RX=D7=GPIO20 ?

**Reply by ReKST-117** [reply]
Hi, you are correct. From the pin out diagram of XIAO esp32c3. D6=TX=GPIO21; D7=RX=GPIO20. So the correct comment should be:   mySerial.begin(115200, SERIAL_8N1, 7, 6); // RX=D7(GPIO20), TX=D6(GPIO21)

**Reply by Elu43** [reply]
Actually after spending 2 days on it I discovered that on this expansion board, it's using a TP8485E chip to convert the signal. The inputs of this chip are actually physically connected to D4 and D5 on the expansion board. But it's easy to make a mistake looking at the pinout diagram. So for the C3 it needs to be :

`  mySerial.begin(115200, SERIAL_8N1, 7, 6); // RX=D5(GPIO7), TX=D4(GPIO6)`

And for C6 :

`  mySerial.begin(115200, SERIAL_8N1, 23, 22); // RX=D5(GPIO23), TX=D4(GPIO22)`

**Comment by TPPS999** [comment]
Working example for esp32s3

// RS485 on UART1 - TESTED & WORKING!
HardwareSerial RS485(1);
#define RS485_RX    6     // GPIO6 - Working configuration!
#define RS485_TX    5     // GPIO5 - Working configuration!
#define RS485_DE_RE 3     // GPIO3 - DE/RE control

void setup() {
    // Keep USB Serial for debugging  
    Serial.begin(115200);
    delay(1000);
    
    Serial.println("=== XIAO ESP32S3 RS485 Test ===");
    Serial.println("USB Serial: Debug messages");
    Serial.println("UART1: RS485 communication");
    
    // Initialize RS485 on UART1
    RS485.begin(9600, SERIAL_8N1, RS485_RX, RS485_TX);
    pinMode(RS485_DE_RE, OUTPUT);
    digitalWrite(RS485_DE_RE, LOW);  // Start in RX mode
    
    Serial.println("RS485 initialized successfully!");
}

void loop() {
    // Send RS485 message
    digitalWrite(RS485_DE_RE, HIGH);  // TX mode
    RS485.println("Hello RS485!");
    RS485.flush();
    digitalWrite(RS485_DE_RE, LOW);   // RX mode
    
    // Check for response
    if (RS485.available()) {
        String response = RS485.readString();
        Serial.println("RS485 Response: " + response);
    }
    
    delay(1000);
}

**Comment by Daniel22ch** [comment]
With this board, the system can be energized by 12 VDC??

**Reply by Lukilyy** [reply]
Hello, regarding your question. This board is designed with a voltage reduction circuit based on the TPS54302 chip, which can safely convert 12V voltage into the 5V required by the system. All you need to do is connect a 12-volt power supply to the 12V power input position of XIAO.

<img width="1104" height="1050" alt="image" src="https://github.com/user-attachments/assets/7a39aec4-a6af-4959-9977-ed92072b5544" />


<img width="2321" height="1599" alt="image" src="https://github.com/user-attachments/assets/e056e158-0ad3-40ce-8373-6eef0b30c66f" />

**Comment by yd65wgs2v7-creator** [comment]
Hi, does anyone know how to map for XIAO c5?

**Reply by Lukilyy** [reply]
Hi,

For XIAO ESP32-C5 with the XIAO RS485 Expansion Board, please continue using the same XIAO header pins as shown in the Wiki example: D4 and D5.

The difference is that the GPIO mapping on XIAO ESP32-C5 is different from XIAO ESP32-C3:

XIAO ESP32-C3:

* D4 = GPIO6
* D5 = GPIO7

XIAO ESP32-C5:

* D4 = GPIO23
* D5 = GPIO24

So on XIAO ESP32-C5, please modify the UART initialization to:

```cpp id="8m9p4v"
mySerial.begin(115200, SERIAL_8N1, 24, 23);
// RX=D5(GPIO24), TX=D4(GPIO23)
```

The enable pin can still remain:

```cpp id="1k3v0n"
#define enable_pin D2
```

In other words, you only need to update the UART RX/TX pin mapping according to the XIAO ESP32-C5 GPIO definition.

**Comment by tommyloon** [comment]
I bought three of these boards together with Seeed Studio XIAO ESP32-C6 modules, but unfortunately I can't get any of them to work.

For testing, I'm trying to read data from an SDM120M energy meter. I know the meter is working because I can read it successfully using an RS485-to-USB adapter.

I also tested the board with a Wemos D1 Mini (ESP8266), but I get exactly the same result, so the issue doesn't appear to be specific to the ESP32-C6.

This is my current ESPHome configuration:

uart:
  id: modbus_uart
  tx_pin: GPIO22
  rx_pin: GPIO23
  baud_rate: 2400
  stop_bits: 1
  flow_control_pin: GPIO2

sensor:
  - platform: sdm_meter
    phase_a:
      active_power:
        name: "SDM230M Power"

Can anyone see what I might be doing wrong?

I've already tried a second board with another XIAO ESP32-C6, but the result is exactly the same. The boards boot normally, connect to Wi-Fi without any issues, but all sensor values remain Unknown.

Has anyone successfully used this board with ESPHome and Modbus/RS485, or can anyone point me in the right direction?

**Reply by FaiyuetCik** [reply]
You can try modifying it like this:
1. Under the uart: section, delete the line flow_control_pin: GPIO2
2. Add a new `modbus:` section between uart: and sensor:, and place flow_control_pin: GPIO2 in this section.

After the revision, it looks like this:

```yaml
uart:
  tx_pin: GPIO22
  rx_pin: GPIO23
  baud_rate: 2400
  stop_bits: 1
  parity: NONE

modbus:
  flow_control_pin: GPIO2

sensor:
  - platform: sdm_meter
    phase_a:
      active_power:
        name: "SDM230M Power"
```

Why can the RS485-USB adapter read data, but the expansion board cannot?

The RS485-USB adapter is equipped with an internal automatic direction control circuit, and the transmission-reception switching is automatically completed by hardware without additional control.

The TP8485E chip on the XIAO RS485 expansion board is half-duplex, and its transmit/receive direction must be manually switched via the D2 pin: high level = transmit, low level = receive. When you write flow_control_pin under the uart: section, ESPHome will use the ESP32 hardware RS485 mode. This mode is not supported by the ESP8266, and on the ESP32-C6 it can also easily be bypassed by the implicitly created modbus component, so D2 is not actually controlled and the Modbus requests never get sent out.

After switching to modbus. flow_control_pin, ESPHome will automatically pull D2 high before each transmission and pull it low once the transmission is completed.

Give it a try, it should work fine.
PS: You can also try checking the signal levels with an oscilloscope to pinpoint where the issue is.

**Reply by tommyloon** [reply]
Thank you, changing the flow_control_pin to modbus: is working!
